rclcpp_action/test/test_actions.cpp

```diff
@@ -0,0 +1,286 @@
+        actions_test_data_t{ false, true,  true  },
+        actions_test_data_t{ true,  false, false },
+        actions_test_data_t{ true,  false, true  },
+        actions_test_data_t{ true,  true,  false },
+        actions_test_data_t{ true,  true,  true  }
+    ));
+
+TEST_P(ActionsTest, SucceedGoal)
+{
+    executor->add_node(server_node);
+    executor->add_node(client_node);
+
+    bool server_available = action_client->wait_for_action_server(std::chrono::seconds(1));
+    ASSERT_TRUE(server_available);
+
+    auto goal_handle_future = action_client->async_send_goal(goal_msg, send_goal_options);
+    auto accepted_response_wait = goal_handle_future.wait_for(std::chrono::seconds(5));
+    ASSERT_TRUE(accepted_response_wait == std::future_status::ready) << "Goal was rejected by server";
+
+    auto goal_handle = goal_handle_future.get();
+    ASSERT_TRUE(goal_handle != nullptr) << "Invalid goal";
+
+    auto result_future = action_client->async_get_result(goal_handle);
+
+    test_info->succeed_goal();
+
+    auto result_wait = result_future.wait_for(std::chrono::seconds(5));
+    ASSERT_TRUE(result_wait == std::future_status::ready) << "Goal not completed on time";
+
+    auto wrapped_result = result_future.get();
+    EXPECT_EQ(wrapped_result.code, rclcpp_action::ResultCode::SUCCEEDED);
+    EXPECT_TRUE(test_info->result_is_correct(
+        wrapped_result.result->sequence, rclcpp_action::ResultCode::SUCCEEDED));
+}
+
+TEST_P(ActionsTest, CancelGoal)
+{
+    executor->add_node(server_node);
+    executor->add_node(client_node);
+    send_goal_options.result_callback = nullptr;
+
+    bool server_available = action_client->wait_for_action_server(std::chrono::seconds(1));
+    ASSERT_TRUE(server_available);
+
+    auto goal_handle_future = action_client->async_send_goal(goal_msg, send_goal_options);
+    auto accepted_response_wait = goal_handle_future.wait_for(std::chrono::seconds(5));
+    ASSERT_TRUE(accepted_response_wait == std::future_status::ready) << "Goal was rejected by server";
+    auto goal_handle = goal_handle_future.get();
+    ASSERT_TRUE(goal_handle != nullptr) << "Invalid goal";
+
+    auto cancel_result_future = action_client->async_cancel_goal(goal_handle);
+    auto cancel_response_wait = cancel_result_future.wait_for(std::chrono::seconds(5));
+    ASSERT_TRUE(cancel_response_wait == std::future_status::ready) << "Cancel response not on time";
+    auto cancel_result = cancel_result_future.get();
+    ASSERT_TRUE(cancel_result != nullptr) << "Invalid cancel result";
+    EXPECT_NE(cancel_result->return_code, action_msgs::srv::CancelGoal::Response::ERROR_REJECTED);
+
+    std::this_thread::sleep_for(std::chrono::milliseconds(10));
+
+    auto result_future = action_client->async_get_result(goal_handle);
+    auto result_response_wait = result_future.wait_for(std::chrono::seconds(5));
+    ASSERT_TRUE(result_response_wait == std::future_status::ready) << "Cancel result response not on time";
```

Comment:
include `<future>` and probably linters will fail here, this line has more then 100 characters